src/gei/Commands/GenerateArchiveCommand.cs

```diff
@@ -0,0 +1,150 @@
+            }
+
+            var githubApi = noSslVerify ? _sourceGithubApiFactory.CreateClientNoSSL() : _sourceGithubApiFactory.Create();
+
+            var repositories = new string[] { sourceRepo };
+
+            var gitDataOptions = new
+            {
+                repositories,
+                exclude_metadata = true
+            };
+            var gitDataArchiveId = await githubApi.StartArchiveGeneration(ghesUrl, githubSourceOrg, gitDataOptions);
+
+            _log.LogInformation($"Archive generation of git data started with id: {gitDataArchiveId}");
+
+            var metadataOptions = new
+            {
+                repositories,
+                exclude_git_data = true,
+                exclude_releases = true,
+                exclude_owner_projects = true
+            };
+            var metadataArchiveId = await githubApi.StartArchiveGeneration(ghesUrl, githubSourceOrg, metadataOptions);
+
+            _log.LogInformation($"Archive generation of metadata started with id: {metadataArchiveId}");
+
+
+            var ids = new int[] { gitDataArchiveId, metadataArchiveId };
+
+            var archiveIdsWithFinished = new Dictionary<int, bool>()
+            {
+                { gitDataArchiveId, false },
+                { metadataArchiveId, false }
+            };
+
+            var currFinished = 0;
+            var total = ids.Length;
+            var timeOut = DateTime.Now.AddHours(_timeoutInHours);
+
+            while (currFinished < total && DateTime.Now < timeOut)
+            {
+                foreach (var pair in archiveIdsWithFinished)
+                {
+                    if (pair.Value)
+                    {
+                        continue;
+                    }
+                    var id = pair.Key;
+                    var archiveStatus = await githubApi.GetArchiveMigrationStatus(ghesUrl, githubSourceOrg, id);
+                    var stringStatus = GithubEnums.ArchiveMigrationStatusToString(archiveStatus);
+
+                    _log.LogInformation($"Waiting for archive with id {id} generation to finish. Current status: {stringStatus}");
+
+                    if (archiveStatus == GithubEnums.ArchiveMigrationStatus.Exported)
+                    {
+                        archiveIdsWithFinished[id] = true;
+                        currFinished++;
+                    }
+                    else if (archiveStatus == GithubEnums.ArchiveMigrationStatus.Failed)
+                    {
+                        _log.LogError($"Archive generation failed with id: {id}");
+                        return;
+                    }
+
+                }
+
+                await Task.Delay(_delayInMS);
+            }
+
+            foreach (var id in ids)
+            {
+                var urlLocation = await githubApi.GetArchiveMigrationUrl(ghesUrl, githubSourceOrg, id);
+                _log.LogInformation($"Archive dowload url: {urlLocation}");
+            }
+        }
```

Comment:
I took a crack at refactoring this a bit:

```suggestion
            var gitArchiveUrl = await WaitForArchiveGeneration(gitDataArchiveId);
            _log.LogInformation($"Archive(git) download url: {gitArchiveUrl}");

            var metadataArchiveUrl = await WaitForArchiveGeneration(metadataArchiveId);
            _log.LogInformation($"Archive(metadata) download url: {metadataArchiveUrl}");
        }

        public async Task<string> WaitForArchiveGeneration(GithubApi githubApi, string ghesUrl, string githubSourceOrg, int archiveId)
        {
            var timeOut = DateTime.Now.AddHours(_timeoutInHours);

            while (DateTime.Now < timeOut)
            {
                var archiveStatus = await githubApi.GetArchiveMigrationStatus(ghesUrl, githubSourceOrg, archiveId);
                _log.LogInformation($"Waiting for archive with id {archiveId} generation to finish. Current status: {archiveStatus}");

                if (archiveStatus == ArchiveMigrationStatus.Exported)
                {
                    return await githubApi.GetArchiveMigrationUrl(ghesUrl, githubSourceOrg, archiveId);
                }

                if (archiveStatus == ArchiveMigrationStatus.Failed)
                {
                    throw new OctoshiftCliException($"Archive generation failed with id: {archiveId}");
                }

                await Task.Delay(_delayInMS);
            }

            throw new TimeoutException($"Archive generation timed out after {_timeoutInHours} hours");
        }
```

Reply:
Sounds good, I can adopt this refactor